spec/app/ics-029-fee-payment/README.md

```diff
@@ -122,32 +149,24 @@ function timeoutPacket(
   proofHeight: Height,
   nextSequenceRecv: Maybe<uint64>,
   relayer: string): Packet {
-    // ...
     // get the timeout relayer from function arguments
     // and pay timeout fee.
     // NOTE: Fee may be zero
     PayTimeoutFee(packet, relayer)
 }
 ```
 
-The fee module should also expose the following queries so that relayers may query their expected fee:
-
-```typescript
-// Gets the fee expected for submitting ReceivePacket msg for this packet
-function GetReceiveFee(packet) Fee
+#### Reasoning
 
-// Gets the fee expected for submitting AcknowledgePacket msg for this packet
-function GetAckFee(packet) Fee
+This proposal satisfies the desired properties. All parts of the packet flow (receive/acknowledge/timeout) can be properly incentivized and rewarded. The protocol does not specify the relayer beforehand, thus the incentivization is permissionless. The escrowing and distribution of funds is completely handled on source chain, thus there is no need for additional IBC packets or the use of ICS-20 in the fee protocol. The fee protocol only assumes existence of fungible tokens on the source chain. Using the connection version, the protocol enables opt-in incentivization and backwards compatibility. The fee module can implement arbitrary custom logic so long as it respects the callback interfaces that IBC expects.
 
-// Gets the fee expected for submitting TimeoutPacket msg for this packet
-function GetTimeoutFee(packet) Fee
-```
+##### Correctness
 
-Since different chains may have different representations for fungible tokens and this information is not being sent to other chains; this ICS does not specify a particular representation for the `Fee`. Each chain may choose its own representation, it is incumbent on relayers to interpret the Fee correctly.
+The fee module is responsible for correctly escrowing and distributing funds to the provided relayers. It is IBC's responsibility to provide the fee module with the correct relayers. The ack and timeout relayers are trivially retrievable since they are the senders of the acknowledgment and timeout message.
 
-#### Reasoning
+The receive relayer submits the message to the counterparty chain. Thus the counterparty chain must communicate the knowledge of who relayed the receive packet to the source chain using the acknowledgement. The address that is sent back **must** be the address of the forward relayer on the source chain.
```

Comment:
This seems like the trickiest part to get right and I would expand on that here.

I think this works, but how do we handle all possible errors?
What if eg. the IbcReceivePacket doesn't have a `payOnSender` fields?
What if the *forward relayer* places an invalid address there? <- this case cannot be proven on the destination chain.

There is some level of "best effort" here. We should be clear under what conditions we are unable to provide that info. "Only if the forward relayer provides correct info" may be a valid caveat.

Reply:
It would be great if we can enforce that IBC processing ensures that if forward relayer passes invalid address, it will not pass validation check on source chain. In my view, that's the most important invariant we need to ensure.

Reply:
Yes, this is definitely something I should specify in this doc. I think we can only go with "best efforts" approach here, because we do not want to fail packet processing if relayer makes a mistake.

As far as I can see there are two scenarios where this can happen.

It must either be caused by the forward relayer making a mistake => I see this as not as issue. They can fix their code to do it correctly.
Or it must be the case that the counterparty chain is implemented incorrectly => In this case, we would not want to prevent acknowledgements from getting processed. We would just 

If we go with a "best effort" approach which is where I'm leaning, then relayers who incorrectly set their address simply don't get paid until they fix the issue.
An incorrect counterparty implementation will initially burn some forward relayers, but they can learn quickly and the channel becomes a still functioning but unincentivized channel in forward direction until a fix is made.

In the case where we return error
A packet will effectively fail if forward address is incorrect. This is bad because it's already processed on counterparty and acknowledgment has been written. Thus we cannot retry with a better relayer, and acknowledgement can never be processed.
In case of incorrect implementation, acknowledgements can never be properly processed on source chain.

I think we should go with a "best efforts" approach here, and clearly define when we will not pay forward fee so that relayers can adapt if they are not getting their expected fees for some reason.

Reply:
Good reasoning.
This should definitely be made explicit in this doc. (Ideally with the reasoning, like you just did above)